### PR TITLE
Roachtest SSH Retries

### LINKED ISSUE #73542: roachtest: auto retry operations when failing due to ssh-related issues
When a roachprod operation fails due to an ssh error, attempting it again will likely work. This will help in making roachtests less prone to failure due to issues unrelated to the tests such as intermittent SSH errors.

Epic CRDB-10428
Jira issue: CRDB-11630

### PR DESCRIPTION
**Pull request by smg260**
roachtest:  This change introduces a default retry when encountering an error of `255` (SSH). It is transparent to callers. If desired, it would be relatively simple to expose this via the `cluster` interface and allow callers to specify a retry handler/options on a per command basis, perhaps something like:

``` 
// retry if we get an exit code of 12
retryOpts { 
	opt: retry.Options { .. } , 
	shouldRetryFn: func(res RunResultDetails) bool { return res.RemoteExitStatus == 12 }
}
c.RunE(ctx, nodes, retryOpts, "my_cmd")
```
but this PR does not introduce that.

Today there appears the concept of a "roachprod" and a "command error", the former of which denotes an issue in roachprod handling code, the latter representing an error executing a command over SSH. This PR preserves this behaviour and introduces an updated function signature from `f(i int) ([]byte, error)` to `f(i int) RunResultDetails, error`. `RunResultDetails` has been expanded to capture information about the execution of a command, such as `stderr/our, exit code, error, attempt number`. Any roachprod error will result in an `error` being returned, and set in `RunResultDetails.Err`. Any command error would be only set in `RunResultDetails.Err` with the returned error being nil. This allows callers to differentiate between a roachprod and a command error - something which existing code depends on.

Retry handling code can use a function's returned `RunResultDetails` to determine whether a retry should take place. This PR has default retry handling on `RunResultDetails.RemoteExitStatus == 255`.

Notable changes
- Retry on `255` exit code for any function executed via `ParallelE`
- Consolidated `Run/RunWithDetails` (significant duplication of code)
- Classify exit error at the session.go level so that it is available for all commands and to all callers
- Remove `"echo $?"` option to print exit code at end of cmd. It is not clear why this was done. (Perhaps for visibility in logs?) Accompanying function to parse out the status has also been removed. The exit code is still accessible via the `RunResultDetails`
- Update roachtests to check for SSH marker error instead of checking for `!install.NonZeroExitCode`

Resolves: https://github.com/cockroachdb/cockroach/issues/73542
Release note: None
Epic: [CRDB-21386](https://cockroachlabs.atlassian.net/browse/CRDB-21386)

### CODE REVIEW

**Comment by smg260** [comment]
Second part updates function signature from `f(i int) ([]byte, error)` to `f(i int) RunResultDetails, error` which allows us to act on the data such as exit codes, stdout, stderr, combinedoutput. In this case, we retry based on the `error` or `exit code` in the `RunResultDetails`

**Review by tbg — COMMENTED** [review]
Looks good! I didn't review in detail. Left some inline comments - apologies if some of these are things you're already doing that I simply missed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-:-NGGsiLW8XGqtXJnBph8:b-l7r959)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/django.go` line 190 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGGqSKV5bZI291UQhvd:-NGGqSKV5bZI291UQhve:bflfns2) ([raw file](https://github.com/cockroachdb/cockroach/blob/14d1b7c1becbac2a0171e908810614086fb95987/pkg/cmd/roachtest/tests/django.go#L190)):*
> ```Go
> 			result, err := c.RunWithDetailsSingleNode(ctx, t.L(), node, fmt.Sprintf(djangoRunTestCmd, testName))
> 
> 			if errors.Is(err, rperrors.ErrSSH255) {
> ```

Why do we need this now? I would hope that your average roachtest wouldn't ever need to worry about SSH errors again, since if one occurs it would be caught by the test harness and the failure would be redirected away from the owning team.

True, something like this wouldn't get captured:

```go
err := c.RunE(/* stuff */)
if err != nil {
  t.Fatalf("doing something stupid with this error: %s", err.Error())
}
```

but honestly I wouldn't worry about that - you've built things so that idiomatic failing as a result of an error will always be captured.
___
*[`pkg/roachprod/install/cluster_synced.go` line 125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGGrbHi4ug1J7KhDegX:-NGGrbHi4ug1J7KhDegY:bb07p7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/14d1b7c1becbac2a0171e908810614086fb95987/pkg/roachprod/install/cluster_synced.go#L125)):*
> ```Go
> // If the `shouldRetryFn` is not specified (nil), then no retries will be
> // performed.
> func runWithMaybeRetry(
> ```

Retries are only admissible if what we're invoking is idempotent, which is mostly not the case in practice (I think). I think it's a sane default policy to hope for the best (i.e. it's idempotent or the first execution didn't actually take place) but let's make sure the final error, if one results, is decorated with the eventual fact of a retry having been carried out. We don't want devs wondering why they get a particular error not realizing that there was an internal retry they didn't have on the radar. I see that you're logging, but also makes sense to wrap the reported error if it resulted from a retry.

We should also consider making the retry behavior optional for callers that want to opt out of it for a specific invocation.

**Review by smg260 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/django.go` line 190 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGGqSKV5bZI291UQhvd:-NGItv6g9nI1XpgyMUwI:bykuxaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/14d1b7c1becbac2a0171e908810614086fb95987/pkg/cmd/roachtest/tests/django.go#L190)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Why do we need this now? I would hope that your average roachtest wouldn't ever need to worry about SSH errors again, since if one occurs it would be caught by the test harness and the failure would be redirected away from the owning team.

True, something like this wouldn't get captured:

```go
err := c.RunE(/* stuff */)
if err != nil {
  t.Fatalf("doing something stupid with this error: %s", err.Error())
}
```

but honestly I wouldn't worry about that - you've built things so that idiomatic failing as a result of an error will always be captured.
</blockquote></details>

I've added a paraphrased version of the previous explanatory comment. Basically, `c.RunWithDetailsSingleNode` may encounter an expected command error, in which case the test can proceed.  We want to mark abort on an unexpected roachprod or SSH error
___
*[`pkg/roachprod/install/cluster_synced.go` line 125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGGrbHi4ug1J7KhDegX:-NGIuS7h1pGyBfGgJkfV:bc1bule) ([raw file](https://github.com/cockroachdb/cockroach/blob/14d1b7c1becbac2a0171e908810614086fb95987/pkg/roachprod/install/cluster_synced.go#L125)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Retries are only admissible if what we're invoking is idempotent, which is mostly not the case in practice (I think). I think it's a sane default policy to hope for the best (i.e. it's idempotent or the first execution didn't actually take place) but let's make sure the final error, if one results, is decorated with the eventual fact of a retry having been carried out. We don't want devs wondering why they get a particular error not realizing that there was an internal retry they didn't have on the radar. I see that you're logging, but also makes sense to wrap the reported error if it resulted from a retry.

We should also consider making the retry behavior optional for callers that want to opt out of it for a specific invocation.
</blockquote></details>

Now wrapping in a new error, and marking it with a retry reference error.

Agreed on optionality - internally it is optional, externally it could be optional as per example in my first comment. If it makes sense, I will add as additional functionality in a follow up.

**Review by srosenberg — COMMENTED** [review]
Looks good! I'll have another pass tomorrow with fresh eyes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-:-NGQQ9o-2rlUF63w86OK:b-1wuwgk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @smg260)

___
*[`pkg/roachprod/install/cluster_synced.go` line 2220 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGQPuHR0t_BnvkDKWDb:-NGQPuHR0t_BnvkDKWDc:buoy8ri) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L2220)):*
> ```Go
> ) error {
> 	// FIXME: according to ParallelE, if err != nil, then failed contains the failures for each node
> 	// however right now, we get failures when the error == nil, and report on them
> ```

I am only seeing two returns out of `ParallelE`,

```
	if len(failed) > 0 {
		return failed, errors.New("one or more parallel execution failure")
	}
	return nil, nil
```

**Review by renatolabs — COMMENTED** [review]
Nice work! :clap: Left a few comments, most of them optional.

> Reviewed 12 of 16 files at r1, 1 of 5 files at r5, 2 of 4 files at r6, 4 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-:-NGS-J4t6z6PfpyBFS2T:b-9m9dpo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @smg260)

___
*[`pkg/roachprod/install/cluster_synced.go` line 631 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRpbtEBsQ--bo87iRB:-NGRpbtFEqJ4q5OFinTg:b97279t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L631)):*
> ```Go
> // Error vs result
> // An error is an unexpected state within roachprod
> // A result is the output of running a cmd (could be interpreted as an error)
> ```

Nit: I generally prefer (and seems to be the pattern in other function comments in the codebase as well) to see documentation in the form of full sentences. The documentation here reads more as a note and it took me a while to connect what's here with the function itself.
___
*[`pkg/roachprod/install/cluster_synced.go` line 699 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRq4uS4M3WfCIrFn_I:-NGRq4uS4M3WfCIrFn_J:bng4sg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L699)):*
> ```Go
> //
> // stdout: Where stdout messages are written
> // stderr: Where stderr messages areå written
> ```

:mag: 
___
*[`pkg/roachprod/install/cluster_synced.go` line 715 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRqCLd1L5xAVRCgKA0:-NGRqCLd1L5xAVRCgKA1:bcuvn21) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L715)):*
> ```Go
> 	results := make([]RunResultDetails, len(nodes))
> 	if _, err := c.ParallelE(l, display, len(nodes), 0, func(i int) (RunResultDetails, error) {
> 		//err is a non command failure
> ```

Nit lvl 100: space between `//` and comment.
___
*[`pkg/roachprod/install/cluster_synced.go` line 726 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRr7pX-AfJFz4jmHYR:-NGRr7pX-AfJFz4jmHYS:b-s9zsuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L726)):*
> ```Go
> }
> 
> func selectError(results []RunResultDetails, stream bool, stdout io.Writer) error {
> ```

Do we need this separate function? It's unexpected, to me, that `selectError` is the function that prints errors to stdout.
___
*[`pkg/roachprod/install/cluster_synced.go` line 752 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRrOZa59B2V3xJjnqd:-NGRrOZa59B2V3xJjnqe:bz5dh7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L752)):*
> ```Go
> 		return result, err
> 	}); err != nil {
> 		l.Printf("%v", err)
> ```

Is this intended? If so, we should add more context.
___
*[`pkg/roachprod/install/cluster_synced.go` line 828 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRrkAS69CxJv5p8B6l:-NGRrkAS69CxJv5p8B6m:bcrupdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L828)):*
> ```Go
> }
> 
> func getRunResult(f func() error, res *RunResultDetails) {
> ```

Nits: I find these two functions not very idiomatic. First, taking a function as parameter seems overly abstract, in reality there's just one (or a couple) function that should be called that makes this function meaningful. Why not take in `(err)` as parameter and let the caller call the function?

Also, do we need to reuse the `RunResultDetails` struct here? It's more natural to me if these were immutable and returned a new instance when called instead.
___
*[`pkg/roachprod/install/cluster_synced.go` line 741 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGS-2rE-TmSSQziZEO3:-NGS-2rE-TmSSQziZEO4:b-wnnw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2c44fffacbdfd6c8829f23c8a61a528e5350fc3/pkg/roachprod/install/cluster_synced.go#L741)):*
> ```Go
> }
> 
> // selectError returns the error from the RunResultDetails with the highest RemoteExitStatus
> ```

Maybe a comment as to why this is a good way to prioritize?

**Comment by renatolabs** [comment]
I kicked the tires a bit, and the retries seem to work well in a basic roachtest I wrote. I have a few questions that I think we should have an answer for before merging:

* I don't see intermediary errors in the test logs; I only see the final `error persisted after 3 retries`.
* I suspect a fair percentage of SSH-related problems (255) we see happen for long running processes (such as `workload`). For example, in [this failure](https://github.com/cockroachdb/cockroach/issues/90695#issuecomment-1295843128), the workload ran for 16 minutes before the process exited with exit code 255. What's the impact of this change in failures like that?
* Do we understand the conditions in which a [test time out](https://github.com/cockroachdb/cockroach/issues/90695#issuecomment-1297834753) can lead to 255 exit status? Retrying a command that timed out sounds bad.

File: pkg/cmd/roachtest/tests/BUILD.bazel
Comment (renatolabs) [inline pkg/cmd/roachtest/tests/BUILD.bazel:50]: 👀

**Review by smg260 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/roachprod/install/cluster_synced.go` line 726 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRr7pX-AfJFz4jmHYR:-NGT0vz8D-rsGCJQ9iI6:b-h5w6hc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L726)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Do we need this separate function? It's unexpected, to me, that `selectError` is the function that prints errors to stdout.
</blockquote></details>

I think its a distinct enough unit of work to be separated. Having errors printed out is unexpected, but both the printing and selecting of an error require a full iteration of the slice.
___
*[`pkg/roachprod/install/cluster_synced.go` line 752 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRrOZa59B2V3xJjnqd:-NGT1mU-5TNvd6jvttF_:bmxldg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L752)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Is this intended? If so, we should add more context.
</blockquote></details>

It was there to satisfy a linter since we are ignoring both outputs of the `c.ParallelE`. Hopefully what's there now is ok.
___
*[`pkg/roachprod/install/cluster_synced.go` line 828 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRrkAS69CxJv5p8B6l:-NGT21HZ5BPeTx4BAWXh:b-c3wrgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L828)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nits: I find these two functions not very idiomatic. First, taking a function as parameter seems overly abstract, in reality there's just one (or a couple) function that should be called that makes this function meaningful. Why not take in `(err)` as parameter and let the caller call the function?

Also, do we need to reuse the `RunResultDetails` struct here? It's more natural to me if these were immutable and returned a new instance when called instead.
</blockquote></details>

Done. I'll defer to your experience regarding idiomatic golang code.  Still a noob in that regard :) The intent of the func param initially was to show that this function was merely wrapping/enriching the original call.
___
*[`pkg/roachprod/install/cluster_synced.go` line 2220 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGQPuHR0t_BnvkDKWDb:-NGT3bhLAq7E7MfvVjXR:balvor6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L2220)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

I am only seeing two returns out of `ParallelE`,

```
	if len(failed) > 0 {
		return failed, errors.New("one or more parallel execution failure")
	}
	return nil, nil
```
</blockquote></details>

Might have been looking at an old revision. I removed my comment - the confusion stemmed from me these function comments .
```
// ParallelE runs the given function in parallel across the given
// nodes, returning an error if function returns an error.
...
// If err is non-nil, the slice of ParallelResults will contain the
// results from any of the failed invocations.
```

It implies that if any node returns any error,  then `ParallelE` will return an error. It actually will only do that for a roachprod error (as opposed to an error over SSH)

**Review by smg260 — COMMENTED** [review]
Thanks - I also tested with a simple roachtest. 

- You should see all errors in the failure.log. I've since also implemented `CombineErrors` to preserve previous failures
- Failures for long running processes will retry, but ultimately will still be bound by the overall timeout configured for the test. 
- I could not reproduce that case, but like above, after a timeout, SIGQUIT and then SIGKILL will be sent to the process. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-:-NGT4sTBDya9L8ARWpvn:b-1zgj96)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

**Review by smg260 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/BUILD.bazel` line 50 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NEvKGARersNwrpd18qi-r19-50:-NGWyNbl0OjjA1ABVjFv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce1dd399b522eceaf4dc106778b11dd8aa1b13f4/pkg/cmd/roachtest/tests/BUILD.bazel#L50)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

👀 
</blockquote></details>

Done.
___
*[`pkg/roachprod/install/cluster_synced.go` line 699 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRq4uS4M3WfCIrFn_I:-NGWyPE19v0QDF52FhyL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L699)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

:mag: 
</blockquote></details>

Done.
___
*[`pkg/roachprod/install/cluster_synced.go` line 741 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGS-2rE-TmSSQziZEO3:-NGWyRWI0QfUgnCWf6_r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2c44fffacbdfd6c8829f23c8a61a528e5350fc3/pkg/roachprod/install/cluster_synced.go#L741)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Maybe a comment as to why this is a good way to prioritize?
</blockquote></details>

Done.

**Review by srosenberg — APPROVED** [review]
:lgtm:

The latest iteration looks good, only minor comments. Improvements look great, but there were also quite a few changes, hence the added caution.
Assuming (a) roachprod-stress didn't reveal any regressions and @renatolabs, @herkolategan feel confident about the changes, let's merge! 

> Reviewed 3 of 16 files at r1, 1 of 5 files at r5, 1 of 4 files at r6, 1 of 13 files at r12, 13 of 13 files at r16, 9 of 9 files at r17, 5 of 5 files at r18, 2 of 2 files at r19, 1 of 3 files at r20.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-:-NGbzExe8O-5aInFdDgn:b9hu2es)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @smg260)

___
*[`pkg/roachprod/install/cluster_synced.go` line 726 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRr7pX-AfJFz4jmHYR:-NGbv9WvDCMXIUFtF4jb:bqlulj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L726)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

I think its a distinct enough unit of work to be separated. Having errors printed out is unexpected, but both the printing and selecting of an error require a full iteration of the slice.
</blockquote></details>

Maybe renaming would denote more clearly what this func is doing? E.g.,

```
// processRunOutput returns the error from the RunResultDetails with a _maximum_ RemoteExitStatus.
// Optionally, if stream==false, each node's combined output is logged to stdout.

func processRunOutput(results []*RunResultDetails, stream bool, stdout io.Writer) error 
```
___
*[`pkg/roachprod/install/cluster_synced.go` line 766 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGbxOFCCFDudpV63Ul9:-NGbxOFCCFDudpV63UlA:b-b0mfgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/78a8c2ddd0bd6daef44f50451cf39d6246f4af14/pkg/roachprod/install/cluster_synced.go#L766)):*
> ```Go
> 
> // selectError returns the error from the RunResultDetails with the highest RemoteExitStatus.
> // This preserves the heuristic in the previously implemented and used `rperrors.SelectPriorityError`.
> ```

I'd remove this comment line since it's not going to be helpful after this PR is merged.

**Review by herkolategan — APPROVED** [review]
Great addition! Looking forward to the increased stability this provides. Left one or two comments where a comment didn't make sense to me.
:lgtm:

> Reviewed 3 of 16 files at r1, 1 of 5 files at r5, 1 of 4 files at r6, 1 of 13 files at r12, 10 of 13 files at r16, 7 of 9 files at r17, 2 of 5 files at r18, 2 of 3 files at r20, 4 of 4 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-:-NGquoSA4OqjwawBLUZ6:b-t1srru)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @renatolabs and @smg260)

___
*[`pkg/cmd/roachtest/tests/gopg.go` line 119 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGqpJ9sE39Wi2764j_6:-NGqpJ9sE39Wi2764j_7:b-6kq0bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3991bd50e8f76e2b6fd93038de1fcdc8293ed3b5/pkg/cmd/roachtest/tests/gopg.go#L119)):*
> ```Go
> 		)
> 
> 		// Fatal for a roachprod err or SSH error. Roachprod error is when result.Err==nil
> ```

I'm struggling to make sense of the wording on the second sentence here, although I might be missing something. Saw it in a few other places, but only mentioning here.
___
*[`pkg/roachprod/install/cluster_synced.go` line 750 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGqrjQ4B2Tn9C6gEvrK:-NGqrjQ4B2Tn9C6gEvrL:b-9tx3ss) ([raw file](https://github.com/cockroachdb/cockroach/blob/3991bd50e8f76e2b6fd93038de1fcdc8293ed3b5/pkg/roachprod/install/cluster_synced.go#L750)):*
> ```Go
> 	results := make([]*RunResultDetails, len(nodes))
> 
> 	// A result is the output of running a exitCode (could be interpreted as an error)
> ```

Nit: an exitCode

**Review by smg260 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90451)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/roachprod/install/cluster_synced.go` line 726 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGRr7pX-AfJFz4jmHYR:-NGr2oeyAWjZz44udJ5z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37cc6740df07d00d30c67b4849be93556cfd22d/pkg/roachprod/install/cluster_synced.go#L726)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Maybe renaming would denote more clearly what this func is doing? E.g.,

```
// processRunOutput returns the error from the RunResultDetails with a _maximum_ RemoteExitStatus.
// Optionally, if stream==false, each node's combined output is logged to stdout.

func processRunOutput(results []*RunResultDetails, stream bool, stdout io.Writer) error 
```
</blockquote></details>

Done.
___
*[`pkg/roachprod/install/cluster_synced.go` line 766 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/90451#-NGbxOFCCFDudpV63Ul9:-NGr2nAg3y6MKf4yMXy4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/78a8c2ddd0bd6daef44f50451cf39d6246f4af14/pkg/roachprod/install/cluster_synced.go#L766)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

I'd remove this comment line since it's not going to be helpful after this PR is merged.
</blockquote></details>

Done.

**Comment by smg260** [comment]
blathers backport 22.1 22.2

**Pull request merged**
